controllers/etcdcluster_controller.go

```diff
@@ -85,52 +168,68 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
-		if !apierrors.IsNotFound(err) {
-			log.Error(err, "unable to fetch EtcdCluster service")
-			return ctrl.Result{}, err
-		}
-		service = headlessServiceForCluster(cluster)
-		if err := r.Create(ctx, service); err != nil {
-			log.Error(err, "unable to create Service", "service", service.Name)
-			return ctrl.Result{}, err
-		}
-		log.V(1).Info("Created Service", "service", service.Name)
-		return ctrl.Result{}, nil
+	// Attempt to dial the etcd cluster, recording the cluster response if we can
+	members := &[]etcdclient.Member{}
+	if memberSlice, err := r.getEtcdMembers(ctx, cluster); err != nil {
+		log.Error(err, "Unable to contact etcd cluster")
+		members = nil
+	} else {
+		members = &memberSlice
 	}
-	log.V(2).Info("Service exists", "service", service.Name)
 
-	peers := &etcdv1alpha1.EtcdPeerList{}
-	if err := r.List(ctx, peers, client.MatchingFields{clusterNameSpecField: cluster.Name}); err != nil {
-		log.Error(err, "unable to list peers")
-		return ctrl.Result{}, err
-	}
-
-	// Assume we can't communicate with the cluster
-	log.V(1).Info("Unable to communicate with etcd cluster")
-	if cluster.Spec.Replicas != nil && int32(len(peers.Items)) < *cluster.Spec.Replicas {
-
-		// Create more peers
-		peerName := nextAvailablePeerName(cluster, peers.Items)
-		log.V(1).Info("Insufficient peers for replicas, adding new peer",
-			"current-peers", len(peers.Items),
-			"desired-peers", cluster.Spec.Replicas,
-			"peer", peerName)
-		peer := peerForCluster(cluster, peerName)
-		if err := r.Create(ctx, peer); err != nil {
-			log.Error(err, "Failed to create peer",
-				"peer", peerName)
-			return ctrl.Result{}, err
-		}
-		return ctrl.Result{}, nil
+	// Perform a reconcile, getting back the desired result, any errors, and a clusterEvent. This is an internal concept
+	// and is not the same as the Kubernetes event, although it is used to make one later.
+	result, clusterEvent, err := r.reconcile(ctx, members, cluster)
+	if err != nil {
+		log.Error(err, "Failed to reconcile")
+	}
+
+	// The update status takes in the cluster definition, and the member list from etcd as of *before we ran reconcile*.
+	// We also get the event, which may contain rich information about what we did (such as the new member name on a
+	// MemberAdded event).
+	err = r.updateStatus(ctx, cluster, members, clusterEvent)
+	if err != nil {
+		log.Error(err, "Failed to update status")
 	}
 
-	return ctrl.Result{}, nil
+	// Finally, the event is used to generate a Kubernetes event by calling `Record` and passing in the recorder.
+	if clusterEvent != nil {
+		clusterEvent.Record(r.Recorder)
+	}
+
+	return result, nil
+}
+
+func (r *EtcdClusterReconciler) getEtcdMembers(ctx context.Context, cluster *etcdv1alpha1.EtcdCluster) ([]etcdclient.Member, error) {
+	etcdConfig := etcdClientConfig(cluster)
+	c, err := etcdclient.New(*etcdConfig)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to etcd: %w", err)
+	}
```

Comment:
perhaps premature optimisation, but wdyt about keeping a pool of etcd clients open in the reconciler? that way we wouldn't need to establish a new client connection on each reconcile

Reply:
Sounds like a nice to have. I think we'll end up implementing somethng similar for watching the etcd API anyway. So probably a future problem. We could open an issue for it?

Reply:
Raised #61